Code/Framework/AzCore/AzCore/std/string/utf8/unchecked.h

```diff
@@ -166,56 +191,114 @@ namespace Utf8::Unchecked
     template <typename u16bit_iterator, typename Utf8Iterator>
-    u16bit_iterator utf8to16(Utf8Iterator start, Utf8Iterator end, u16bit_iterator result)
+    u16bit_iterator utf8to16(Utf8Iterator start, Utf8Iterator end, u16bit_iterator result, size_t resultMaxSize)
     {
         octet_iterator utf8Start(start);
         octet_iterator utf8Last(end);
-        while (utf8Start != utf8Last)
+        while (utf8Start != utf8Last && resultMaxSize > 0)
         {
             AZ::u32 cp = *utf8Start++;
             if (cp > 0xffff)
             {
                 //make a surrogate pair
-                *result++ = static_cast<AZ::u16>((cp >> 10) + Internal::LEAD_OFFSET);
-                *result++ = static_cast<AZ::u16>((cp & 0x3ff) + Internal::TRAIL_SURROGATE_MIN);
+                if (resultMaxSize >= 2)
+                {
+                    *result++ = static_cast<AZ::u16>((cp >> 10) + Internal::LEAD_OFFSET);
+                    *result++ = static_cast<AZ::u16>((cp & 0x3ff) + Internal::TRAIL_SURROGATE_MIN);
+                    resultMaxSize -= 2;
+                }
+                else
+                {
+                    resultMaxSize = 0;
+                }
             }
             else
             {
                 *result++ = static_cast<AZ::u16>(cp);
+                resultMaxSize -= 1;
             }
         }
         return result;
     }
 
     template <typename Utf8Iterator, typename u32bit_iterator>
-    Utf8Iterator utf32to8(u32bit_iterator start, u32bit_iterator end, Utf8Iterator result)
+    Utf8Iterator utf32to8(u32bit_iterator start, u32bit_iterator end, Utf8Iterator result, size_t resultMaxSize)
     {
-        while (start != end)
+        while (start != end && resultMaxSize > 0)
         {
-            octet_iterator<Utf8Iterator>::to_utf8_sequence(*start++, result);
+            octet_iterator<Utf8Iterator>::to_utf8_sequence(*start++, result, resultMaxSize);
         }
 
         return result;
     }
 
     template <typename Utf8Iterator, typename u32bit_iterator>
-    u32bit_iterator utf8to32(Utf8Iterator start, Utf8Iterator end, u32bit_iterator result)
+    u32bit_iterator utf8to32(Utf8Iterator start, Utf8Iterator end, u32bit_iterator result, size_t resultMaxSize)
     {
         octet_iterator utf8Start(start);
         octet_iterator utf8Last(end);
-        while (utf8Start != utf8Last)
+        while (utf8Start != utf8Last && resultMaxSize > 0)
         {
             *result++ = *utf8Start++;
+            --resultMaxSize;
         }
 
         return result;
     }
+
+    static constexpr size_t utf8_codepoint_length(AZ::u32 cp)
+    {
+        if (cp < 0x80)
+        {
+            return 1;
+        }
+        else if (cp < 0x800)
+        {
+            return 2;
+        }
+        else if (cp < 0x10000)
+        {
+            return 3;
+        }
+        return 4;
+    }
+
+    template <typename u16bit_iterator>
+    size_t utf16ToUtf8BytesRequired(u16bit_iterator start, u16bit_iterator end)
```

Comment:
question: Same comment about tests as before for these updated and added functions

Reply:
added

Reply:
Amazing thanks! 😄